net/libfab/libfab.c

```diff
@@ -0,0 +1,2829 @@
+static inline bool libfab_is_verbs(struct m0_fab__tm *tm);
+static int libfab_bulklist_add(struct m0_fab__tm *tm, struct m0_fab__buf *fb,
+				struct m0_fab__active_ep *aep);
+static uint32_t libfab_wr_cnt_get(struct m0_fab__buf *fb);
+static void libfab_bulk_buf_process(struct m0_fab__tm *tm);
+
+/* libfab init and fini() : initialized in motr init */
+M0_INTERNAL int m0_net_libfab_init(void)
+{
+	int result = 0;
+
+	m0_net_xprt_register(&m0_net_libfab_xprt);
+	m0_net_xprt_default_set(&m0_net_libfab_xprt);
+	return M0_RC(result);
+}
+
+M0_INTERNAL void m0_net_libfab_fini(void)
+{
+	m0_net_xprt_deregister(&m0_net_libfab_xprt);
+}
+
+/**
+ * This function decodes the lnet format address and extracts the ip address and
+ * port number from it.
+ * This is also used to allocate unique transfer machine identifiers for LNet
+ * network addresses with wildcard transfer machine identifier (like
+ * "192.168.96.128@tcp1:12345:31:*").
+ */
+static int libfab_ep_addr_decode_lnet(const char *name, char *node,
+				      size_t nodeSize, char *port,
+				      size_t portSize)
+{
+	char     *at = strchr(name, '@');
+	int       nr;
+	unsigned  pid;
+	unsigned  portal;
+	unsigned  portnum;
+	unsigned  tmid;
+	char      *lp = "127.0.0.1";
+
+	if (strncmp(name, "0@lo", 4) == 0) {
+		M0_PRE(nodeSize >= ((strlen(lp)+1)) );
+		memcpy(node, lp, (strlen(lp)+1));
+	} else {
+		if (at == NULL || at - name >= nodeSize)
+			return M0_ERR(-EPROTO);
+
+		M0_PRE(nodeSize >= (at-name)+1);
+		memcpy(node, name, at - name);
+	}
+	if ((at = strchr(at, ':')) == NULL) /* Skip 'tcp...:' bit. */
+		return M0_ERR(-EPROTO);
+	nr = sscanf(at + 1, "%u:%u:%u", &pid, &portal, &tmid);
+	if (nr != 3) {
+		nr = sscanf(at + 1, "%u:%u:*", &pid, &portal);
+		if (nr != 2)
+			return M0_ERR(-EPROTO);
+		for (nr = 0; nr < ARRAY_SIZE(fab_autotm); ++nr) {
+			if (fab_autotm[nr] == 0) {
+				tmid = nr;
+				break;
+			}
+		}
+		if (nr == ARRAY_SIZE(fab_autotm))
+			return M0_ERR(-EADDRNOTAVAIL);
+	}
+	/*
+	* Hard-code LUSTRE_SRV_LNET_PID to avoid dependencies on the Lustre
+	* headers.
+	*/
+	if (pid != 12345)
+		return M0_ERR(-EPROTO);
+	/*
+	* Deterministically combine portal and tmid into a unique 16-bit port
+	* number (greater than 1024). Tricky.
+	*
+	* Port number is, in binary: tttttttttt1ppppp, that is, 10 bits of tmid
+	* (which must be less than 1024), followed by a set bit (guaranteeing
+	* that the port is not reserved), followed by 5 bits of (portal - 30),
+	* so that portal must be in the range 30..61.
+	*
+	if (tmid >= 1024 || (portal - 30) >= 32)
+		return M0_ERR_INFO(-EPROTO,
+			"portal: %u, tmid: %u", portal, tmid);
+	*/
+
+	if ( portal < 30)
+		portal = 30 + portal;
+
+	portnum  = tmid | (1 << 10) | ((portal - 30) << 11);
+	M0_ASSERT(portnum < 65537);
+	sprintf(port, "%d", portnum);
+	fab_autotm[tmid] = 1;
+	return M0_RC(0);
+}
+
+/**
+ * This function decodes the socket format address and extracts the ip address
+ * and port number from it.
```

Comment:
Can you give an example of this format?

Reply:
From the code, I can see the ep string must start with a protocol family name.

Reply:
Example added